File: mtc-model/src/main/scala/org/drugis/mtc/Partition.scala
/*
 * This file is part of drugis.org MTC.
 * MTC is distributed from http://drugis.org/mtc.
 * Copyright (C) 2009-2011 Gert van Valkenhoef.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.drugis.mtc

class Part[M <: Measurement](t1: Treatment, t2: Treatment,
		val studies: Set[Study[M]]) {
	require(!studies.isEmpty)
	require(allStudiesHave(t1))
	require(allStudiesHave(t2))

	val treatments = Set(t1, t2)

	override def equals(other: Any): Boolean = other match {
		case that: Part[M] => {
			that.treatments == treatments && that.studies == studies }
		case _ => false
	}

	override def hashCode: Int = 31 * treatments.hashCode + studies.hashCode

	override def toString: String =
		"Part({" + treatments.mkString(", ") + "} -> {" +
			studies.map(s => s.id).mkString(", ") + "})"

	private def allStudiesHave(t: Treatment) =
		studies.forall(s => s.treatments.contains(t))
}

/**
 * Represents a valid partition.
 * The set of parts is: {{x, x} -> P} OR {{x, y} -> P, {x, y} -> Q} OR
 * the parts have non-equal treatment sets and form a cycle.
 */
class Partition[M <: Measurement](val parts: Set[Part[M]]) {
	require(validPartition)

	def reduce: Partition[M] =
		if (parts.size < 3) this
		else new Partition(reduce(Cycle(asGraph).edgeSeq))

	private def support(e: (Treatment, Treatment)): Set[Study[M]] =
		parts.find(p => p.treatments == Set(e._1, e._2)) match {
			case Some(p) => p.studies
			case None => throw new IllegalStateException()
		}

	private def reduce(l: List[(Treatment, Treatment)]): Set[Part[M]] = {
		val e = l.head
		val r = reduce(l, e._1, e._2, support(e))
		if (support(l.head) == support(l.last) && r.size > 1) {
			mergeOnVertex(r, e._1)
		} else r
	}

	private def mergeOnVertex(r: Set[Part[M]], v: Treatment) = {
		def other(p: Part[M]) = (p.treatments - v).toList(0)
		val ps = r.filter(p => p.treatments.contains(v)).toList
		require(ps.size == 2)
		require(ps(0).studies == ps(1).studies)
		val merged = new Part(other(ps(0)), other(ps(1)), ps(0).studies)
		(r -- ps) + merged
	}

	private def reduce(l: List[(Treatment, Treatment)], t0: Treatment,
		t1: Treatment, s: Set[Study[M]])
	: Set[Part[M]] = l match {
		case Nil => Set(new Part(t0, t1, s))
		case e :: l1 =>
			if (support(e) == s) reduce(l1, t0, e._2, s)
			else Set(new Part(t0, t1, s)) ++ reduce(l1, e._1, e._2, support(e))
	}

	override def equals(other: Any) = other match {
		case that: Partition[M] => that.parts == parts
		case _ => false
	}

	override def hashCode: Int = parts.hashCode

	override def toString: String = "Partition(" + parts.mkString(", ") + ")"

	private def validPartition: Boolean = {
		val partList = parts.toList
		if (parts.size == 0) false
		else if (parts.size == 1) partList(0).treatments.size == 1
		else if (parts.size == 2)
			partList(0).treatments == partList(1).treatments
		else asCycle match {
			case Some(c) => true
			case None => false
		}
	}

	def asCycle: Option[Cycle[Treatment]] =
		try {
			Some(Cycle(asGraph))
		} catch {
			case _ => None
		}

	private def asGraph = new UndirectedGraph(parts.map(asEdge _))

	private def asEdge(p: Part[M]) = {
		val t = p.treatments.toList
		(t(0), t(1))
	}
}

object Partition {
	def apply[M <: Measurement](network: Network[M], cycle: Cycle[Treatment]) 
	:Partition[M] = new Partition(Set[Part[M]]() ++
		cycle.edgeSeq.map(edgePart(network) _))

	private def edgePart[M <: Measurement](network: Network[M])(
		e: (Treatment, Treatment))
	: Part[M] = new Part(e._1, e._2, network.supportingStudies(e))
}

File: mtc-model/src/main/java/org/drugis/mtc/parameterization/Partition.java
/*
 * This file is part of the GeMTC software for MTC model generation and
 * analysis. GeMTC is distributed from http://drugis.org/gemtc.
 * Copyright (C) 2009-2012 Gert van Valkenhoef.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.drugis.mtc.parameterization;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;

import org.drugis.mtc.graph.GraphUtil;
import org.drugis.mtc.model.Treatment;

import edu.uci.ics.jung.graph.UndirectedGraph;
import edu.uci.ics.jung.graph.UndirectedSparseGraph;
import edu.uci.ics.jung.graph.util.Pair;

public class Partition {
	private final Set<Part> d_parts;
	private boolean d_reduced;

	public Partition(Collection<Part> parts) {
		d_parts = new HashSet<Part>(parts);
		if (!validPartition(d_parts)) {
			throw new IllegalArgumentException("Given parts do not form a valid partition");
		}
		d_reduced = false;
	}
	
	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Partition) {
			Partition other = (Partition) obj;
			return other.d_parts.equals(d_parts);
		}
		return false;
	}
	
	@Override
	public int hashCode() {
		return d_parts.hashCode();
	}
	
	@Override
	public String toString() {
		return "Partition{" + d_parts.toString() + "}";
	}
	
	/**
	 * Get an unmodifiable view of the parts in this Partition.
	 */
	public Set<Part> getParts() {
		return Collections.unmodifiableSet(d_parts);
	}

	/**
	 * Reduce this partition so that all the parts in the resulting partition are independent.
	 */
	public Partition reduce() {
		if (d_reduced) {
			return this;
		}
		
		UndirectedGraph<Treatment, Part> graph = buildGraph(d_parts);
		if (graph == null) {
			return this;
		}
		
		// Start at an arbitrary part connected to the least vertex.
		// If we don't start at the least vertex, the result might vary for cycles that reduce to a point.
		Treatment v0 = CompareUtil.findLeast(graph.getVertices(), TreatmentComparator.INSTANCE);
		Part p = graph.getIncidentEdges(v0).iterator().next(); // Starting part
		Pair<Treatment> treatments = new Pair<Treatment>(v0, otherTreatment(p, v0));
		Set<Part> visited = new HashSet<Part>(); // Visited parts
		
		// Go as far to the right as possible
		Treatment tRight = walk(graph, p, treatments.getFirst(), visited);
		if (tRight.equals(treatments.getFirst())) {
			return new Partition(Collections.singleton(new Part(tRight, tRight, p.getStudies())));
		}
		// Go as far to the left as possible
		Treatment tLeft = walk(graph, p, treatments.getSecond(), visited);
		
		// Now start going around the cycle starting at tRight, until we get to tLeft.
		Set<Part> reduced = new HashSet<Part>();
		reduced.add(new Part(tRight, tLeft, p.getStudies()));		
		do {
			p = nextPart(graph, visited, tRight);
			Treatment tNext = walk(graph, p, tRight, visited);
			reduced.add(new Part(tRight, tNext, p.getStudies()));
			tRight = tNext;
		} while (!tRight.equals(tLeft));
		
		Partition partition = new Partition(reduced);
		partition.d_reduced = true;
		return partition;
	}
	
	public List<Treatment> asCycle() {
		UndirectedGraph<Treatment, Part> graph = buildGraph(d_parts);
		
		List<Treatment> cycle = new ArrayList<Treatment>(d_parts.size() + 1);
		Part p0 = graph.getEdges().iterator().next();
		Pair<Treatment> treatments = new Pair<Treatment>(p0.getTreatments());
		cycle.addAll(treatments);
		for (int i = 1; i < d_parts.size(); ++i) {
			Part pNext = nextPart(graph, cycle.get(i - 1), cycle.get(i));
			cycle.add(otherTreatment(pNext, cycle.get(i)));
		}
		
		return cycle;
	}

	/**
	 * Walk the partition cycle, starting from the part p0 with treatments {t_0, t_1}, going in the direction t_0 --> t_1.
	 * Stop walking when either t_k == t_0 or when p_k and p_0 have different studies.
	 * @param graph The (simple cycle) graph to walk.
	 * @param p0 The starting edge p_0.
	 * @param t0 The starting vertex t_0.
	 * @param visited (OUTPUT) set of visited edges. This is an output parameter (the set is written to).
	 */
	private Treatment walk(final UndirectedGraph<Treatment, Part> graph, final Part p0, final Treatment t0, Set<Part> visited) {
		visited.add(p0);
		Treatment tPrev = t0; // Previously visited vertex
		Treatment tCurr = otherTreatment(p0, t0); // Current vertex
		Part pNext = nextPart(graph, tPrev, tCurr); // Next edge to consider
		while (p0.getStudies().equals(pNext.getStudies()) && !t0.equals(tCurr)) {
			tPrev = tCurr;
			tCurr = otherTreatment(pNext, tCurr);
			visited.add(pNext); // We now considered pNext
			pNext = nextPart(graph, tPrev, tCurr); // So move on to the next part
		}
		return tCurr;
	}

	/**
	 * Determine whether the given set of parts forms a valid partition.
	 */
	private static boolean validPartition(Set<Part> parts) {
		if (parts.isEmpty()) {
			return false;
		}
		if (parts.size() == 1) {
			return parts.iterator().next().getTreatments().size() == 1;
		}
		if (parts.size() == 2) {
			Iterator<Part> iterator = parts.iterator();
			Part p1 = iterator.next();
			Part p2 = iterator.next();
			return p1.getTreatments().equals(p2.getTreatments()) && p1.getTreatments().size() == 2;
		}
		
		// now try to assemble the parts into a cycle
		UndirectedGraph<Treatment, Part> graph = buildGraph(parts);
		return graph == null ? false : GraphUtil.isSimpleCycle(graph);
	}

	/**
	 * Build a graph in which the given parts are the edges.
	 */
	private static UndirectedGraph<Treatment, Part> buildGraph(Set<Part> parts) {
		UndirectedGraph<Treatment, Part> graph = new UndirectedSparseGraph<Treatment, Part>();
		for (Part p : parts) {
			if (p.getTreatments().size() != 2) { // FIXME: add test for this.
				return null;
			}
			Iterator<Treatment> iterator = p.getTreatments().iterator();
			Treatment t1 = iterator.next();
			Treatment t2 = iterator.next();
			graph.addEdge(p, t1, t2);
		}
		return graph;
	}
	
	/**
	 * Get the treatment of this part that is not equal to the given treatment
	 * @param p Part with exactly 2 treatments.
	 * @param t Treatment not to return.
	 * @return The treatment s \in p, s != t.
	 */
	private static Treatment otherTreatment(Part p, Treatment t) {
		Pair<Treatment> treatments = new Pair<Treatment>(p.getTreatments());
		if (treatments.getFirst().equals(t)) {
			return treatments.getSecond();
		} else {
			return treatments.getFirst();
		}
	}

	/**
	 * Get the next part in the cycle, given we were walking from tPrev to tCurr.
	 * @param g The (simple cycle) graph to walk.
	 * @param tPrev The previously visited vertex.
	 * @param tCurr The current vertex.
	 * @return The edge to the next vertex.
	 */
	private static Part nextPart(UndirectedGraph<Treatment, Part> g, Treatment tPrev, Treatment tCurr) {
		Pair<Part> parts = new Pair<Part>(g.getIncidentEdges(tCurr));
		if (parts.getFirst().getTreatments().contains(tPrev)) {
			return parts.getSecond();
		} else {
			return parts.getFirst();
		}
	}
	
	/**
	 * Get the non-visited edge connected to the given vertex.
	 * @param g The (simple cycle) graph to walk
	 * @param visited Set containing all visited edges.
	 * @param tCurr The current vertex. 
	 * @return The edge connected to tCurr that has not been visited.
	 */
	private static Part nextPart(UndirectedGraph<Treatment, Part> g, Set<Part> visited, Treatment tCurr) {
		Pair<Part> parts = new Pair<Part>(g.getIncidentEdges(tCurr));
		if (visited.contains(parts.getFirst())) {
			return parts.getSecond();
		} else {
			return parts.getFirst();
		}
	}
}
